File: Android/dokit/src/main/java/com/didichuxing/doraemonkit/kit/fileexplorer/FileInfoAdapter.java
package com.didichuxing.doraemonkit.kit.fileexplorer;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.widget.recyclerview.AbsRecyclerAdapter;
import com.didichuxing.doraemonkit.widget.recyclerview.AbsViewBinder;
import com.didichuxing.doraemonkit.util.DoKitFileUtil;

/**
 * Created by wanglikun on 2018/10/16.
 */

public class FileInfoAdapter extends AbsRecyclerAdapter<AbsViewBinder<FileInfo>, FileInfo> {
    private OnViewClickListener mOnViewClickListener;
    private OnViewLongClickListener mOnViewLongClickListener;

    public FileInfoAdapter(Context context) {
        super(context);
    }

    @Override
    protected AbsViewBinder<FileInfo> createViewHolder(View view, int viewType) {
        return new FileInfoViewHolder(view);
    }

    @Override
    protected View createView(LayoutInflater inflater, ViewGroup parent, int viewType) {
        return inflater.inflate(R.layout.dk_item_file_info, parent, false);
    }

    public class FileInfoViewHolder extends AbsViewBinder<FileInfo> {
        private TextView mName;
        private ImageView mIcon;
        private ImageView mMoreBtn;

        public FileInfoViewHolder(View view) {
            super(view);
        }

        @Override
        protected void getViews() {
            mName = getView(R.id.name);
            mIcon = getView(R.id.icon);
            mMoreBtn = getView(R.id.more);
        }

        @Override
        public void bind(final FileInfo fileInfo) {
            getView().setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    return mOnViewLongClickListener != null && mOnViewLongClickListener.onViewLongClick(v, fileInfo);
                }
            });
            mName.setText(fileInfo.file.getName());
            if (fileInfo.file.isDirectory()) {
                mIcon.setImageResource(R.mipmap.dk_dir_icon);
                mMoreBtn.setVisibility(View.VISIBLE);
            } else {
                if (DoKitFileUtil.getSuffix(fileInfo.file).equals(DoKitFileUtil.JPG)) {
                    mIcon.setImageResource(R.mipmap.dk_jpg_icon);
                } else if (DoKitFileUtil.getSuffix(fileInfo.file).equals(DoKitFileUtil.TXT)) {
                    mIcon.setImageResource(R.mipmap.dk_txt_icon);
                } else if (DoKitFileUtil.getSuffix(fileInfo.file).equals(DoKitFileUtil.DB)){
                    mIcon.setImageResource(R.mipmap.dk_file_db);
                }else {
                    mIcon.setImageResource(R.mipmap.dk_file_icon);
                }
                mMoreBtn.setVisibility(View.GONE);
            }
        }

        @Override
        protected void onViewClick(View view, FileInfo data) {
            super.onViewClick(view, data);
            if (mOnViewClickListener != null) {
                mOnViewClickListener.onViewClick(view, data);
            }
        }
    }

    public void setOnViewClickListener(OnViewClickListener onViewClickListener) {
        mOnViewClickListener = onViewClickListener;
    }

    public void setOnViewLongClickListener(OnViewLongClickListener onViewLongClickListener) {
        mOnViewLongClickListener = onViewLongClickListener;
    }

    public interface OnViewClickListener {
        void onViewClick(View v, FileInfo fileInfo);
    }

    public interface OnViewLongClickListener {
        boolean onViewLongClick(View v, FileInfo fileInfo);
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/fileexplorer/FileInfoAdapter.kt
package com.didichuxing.doraemonkit.kit.fileexplorer

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.isVisible
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.widget.recyclerview.AbsRecyclerAdapter
import com.didichuxing.doraemonkit.widget.recyclerview.AbsViewBinder

/**
 * @author lostjobs created on 2020/6/14
 */

typealias onItemClick = (View, FileInfo) -> Unit

class FileInfoAdapter(context: Context) :
        AbsRecyclerAdapter<AbsViewBinder<FileInfo>, FileInfo>(context) {

    private var onItemClick: onItemClick? = null
    private var onItemLongClick: onItemClick? = null

    override fun createViewHolder(view: View, viewType: Int): AbsViewBinder<FileInfo> {
        return FileInfoVH(view)
    }

    override fun createView(inflater: LayoutInflater, parent: ViewGroup?, viewType: Int): View {
        return inflater.inflate(R.layout.dk_item_file_info, parent, false)
    }

    fun doOnViewLongClick(onItemLongClick: onItemClick?) {
        this.onItemLongClick = onItemLongClick
    }

    fun doOnViewClick(onItemClick: onItemClick?) {
        this.onItemClick = onItemClick
    }

    inner class FileInfoVH(view: View) : AbsViewBinder<FileInfo>(view) {

        private val mName: TextView = itemView.findViewById(R.id.name)
        private val mIcon: ImageView = itemView.findViewById(R.id.icon)
        private val mMoreBtn: ImageView = itemView.findViewById(R.id.more)

        override fun onViewClick(view: View, data: FileInfo?) {
            super.onViewClick(view, data)
            data?.run {
                onItemClick?.invoke(view, this)
            }
        }

        override fun onViewLongClick(view: View, data: FileInfo?): Boolean {
            val longClick = onItemLongClick
            if (null != data && null != longClick) {
                longClick.invoke(view, data)
                return true
            }
            return super.onViewLongClick(view, data)
        }

        override fun onBind(data: FileInfo, position: Int) {
            mName.text = data.file.name
            if (data.isDirectory) {
                mIcon.setImageResource(R.mipmap.dk_dir_icon)
                mMoreBtn.isVisible = true
            } else {
                mIcon.setImageResource(
                        when {
                            data.isJPG -> {
                                R.mipmap.dk_jpg_icon
                            }
                            data.isDB -> {
                                R.mipmap.dk_file_db
                            }
                            data.isTxt -> {
                                R.mipmap.dk_txt_icon
                            }
                            else -> {
                                R.mipmap.dk_file_icon
                            }
                        }
                )
                mMoreBtn.isVisible = false
            }
        }
    }
}